### ISSUE TITLE
Feature Request: Optimize binary cache exists check

### ISSUE THREAD

**Issue by kgtkr** [NixOS/nix#7221]
**Is your feature request related to a problem? Please describe.**
When running nix build, fetching `*.narinfo` occurs number of drv * config number of `substituters`. This can take a very long time on high latency networks.

**Describe the solution you'd like**
Cache and use the list of cache files that the binary cache server has. For example, if it is an s3 cache, I can use the file list in xml by accessing the root of the url. 

**Describe alternatives you've considered**
Temporarily set `substitute = false`. But this also can't take use of the existing binary cache.

**Additional context**

**Comment by thufschmitt** [comment]
Yes, the binary cache is quite "dumb" by design to allow serving it with a simple file-based http server.

If there are some good backend-specific APIs that can be used to make this faster, I don't see any objection in principle, although these should stay optional.

In the long run, I'd like to see a more performant http cache protocol being designed, but that's quite unlikely to happen any time soon-ish.

**Comment by edolstra** [comment]
That's how the old "manifest" substituter worked (i.e. by downloading files like https://releases.nixos.org/nixpkgs/nixpkgs-0.10/MANIFEST). The problem is that it doesn't scale. Listing the contents of cache.nixos.org would be prohibitively expensive. A bloom filter could be used but even that would be pretty big.

**Comment by nrdxp** [comment]
perhaps we could at least start the download/upload concurrently with the queries (for the paths that have already been checked), to the make the cost less visible?

**Comment by thufschmitt** [comment]
There is already some amount of parallelism going on: Nix batches all the `narinfo` requests at the start of the build, and the download of the `.nar` files is done as part of the build loop, so concurrently with the other builds/substitutions.

The main bottleneck right now I think is that if multiple binary caches are configured, then they will be checked sequentially for each store path.
